File: javascript/export_and_add_rule_to_index.js
// Install the API client: https://www.algolia.com/doc/libraries/sdk/install#javascript
import { algoliasearch } from "algoliasearch";
import "dotenv/config";
import * as fs from "fs";

// Get your Algolia Application ID and (admin) API key from the dashboard: https://www.algolia.com/account/api-keys
// and choose a name for your index. Add these environment variables to a `.env` file:
const ALGOLIA_APP_ID = process.env.ALGOLIA_APP_ID;
const ALGOLIA_API_KEY = process.env.ALGOLIA_API_KEY;
const ALGOLIA_INDEX_NAME = process.env.ALGOLIA_INDEX_NAME;

// Start the API client
// https://www.algolia.com/doc/libraries/sdk/install#test-your-installation
const client = algoliasearch(ALGOLIA_APP_ID, ALGOLIA_API_KEY);

// Create an index name (or connect to it, if an index with the name `ALGOLIA_INDEX_NAME` already exists)
// https://www.algolia.com/doc/libraries/sdk/install#test-your-installation
const indexName = ALGOLIA_INDEX_NAME || "new_index_name";

(async () => {
  // Export Rules for this index
  // https://www.algolia.com/doc/libraries/sdk/methods/search/browse-rules
  await client.browseRules({
    indexName,
    aggregator: (res) =>
      fs.writeFile(
        `${ALGOLIA_INDEX_NAME}_rules.json`,
        JSON.stringify(res.hits),
        (err) => {
          // In case of a error throw err.
          if (err) throw err;
        }
      ),
  });

  console.log(
    `Rules saved as ${ALGOLIA_INDEX_NAME}_rules.json in the current directory`
  );

  // Create a rule
  const rule = {
    objectID: "a-rule-id",
    conditions: [
      {
        pattern: "Jimmie",
        anchoring: "is",
      },
    ],
    consequence: {
      params: {
        filters: "zip_code = 12345",
      },
    },

    // Optionally, to disable the rule change to 'false'
    enabled: true,

    // Optionally, to add valitidy time ranges
    validity: [
      {
        from: Math.floor(Date.now() / 1000),
        until: Math.floor(Date.now() / 1000) + 10 * 24 * 60 * 60,
      },
    ],
  };

  // Save the Rule, and forward it to all replicas of the index.
  const response = await client.saveRule({
    indexName: indexName,
    objectID: "a-rule-id",
    forwardToReplicas: true,
    rule: rule,
  });

  // Log the response
  console.log(response);
})();

File: ruby/export_and_add_rule_to_index.rb
# Install the API client: https://www.algolia.com/doc/libraries/sdk/install#ruby
require 'algolia'
require 'dotenv/load'
require 'json'

# Get your Algolia Application ID and (admin) API key from the dashboard: https://www.algolia.com/account/api-keys
# and choose a name for your index. Add these environment variables to a `.env` file:
ALGOLIA_APP_ID = ENV['ALGOLIA_APP_ID']
ALGOLIA_API_KEY = ENV['ALGOLIA_API_KEY']
ALGOLIA_INDEX_NAME = ENV['ALGOLIA_INDEX_NAME']

# Initialise the client and connect to it
# https://www.algolia.com/doc/libraries/sdk/methods/search#ruby
client = Algolia::SearchClient.create(ALGOLIA_APP_ID, ALGOLIA_API_KEY)

# Create a rule

puts "Creating a rule ..."
objectID = 'a-rule-id'
rule = {
     objectID: objectID,
     conditions: [{
       pattern: 'Jimmie',
       anchoring: 'is'
     }],
     consequence: {
       params: {
         filters: "zip_code = 12345"
       }
     }
   }
   
# Optionally, to disable the rule
rule['enabled'] = false
   
# Optionally, to add validity time ranges
rule['validity'] = [
     {
          from: Time.now.to_i,
          until: (DateTime.now + 10).to_time.to_i,
     }
]

# Save the Rule.
# https://www.algolia.com/doc/libraries/sdk/methods/search/save-rule
client.save_rule(ALGOLIA_INDEX_NAME, objectID, rule)

# Or save the Rule, and forward it to all replicas of the index.
response = client.save_rule(ALGOLIA_INDEX_NAME, objectID, rule, true)

# Wait for asynchronous task to complete
# https://www.algolia.com/doc/libraries/sdk/methods/search/wait-for-task
client.wait_for_task(ALGOLIA_INDEX_NAME, response.task_id)

# Browse rules
# https://www.algolia.com/doc/libraries/sdk/methods/search/browse-rules#ruby
res = client.browse_rules(ALGOLIA_INDEX_NAME)

# Store rules into an array
rules = res.map { |rule| rule } 

if rules.empty?
     puts "No rules are configured for your index yet."
else
     # Create the json file with all rules
     path = "#{ALGOLIA_INDEX_NAME}_rules.json"
     File.open(path, 'wb') do |file|
          file.write(JSON.generate(rules))
     end    
     puts "Rules export completed"
end

